internal/resources/google/sql_database_instance.go

```diff
@@ -0,0 +1,316 @@
+	var costComponents []*schema.CostComponent
+	tier := r.Tier
+
+	availabilityType := "ZONAL"
+	if r.AvailabilityType != "" && !replica {
+		availabilityType = r.AvailabilityType
+	}
+
+	region := r.Region
+	dbVersion := r.DatabaseVersion
+	dbType := sqlInstanceDBVersionToDBType(dbVersion)
+
+	diskType := "PD_SSD"
+	if r.DiskType != "" {
+		diskType = r.DiskType
+	}
+
+	var diskSizeGB int64 = 10
+	if r.DiskSize != 0 {
+		diskSizeGB = r.DiskSize
+	}
+
+	if sqlInstanceTierToResourceGroup(tier) != "" && dbType != SQLServer {
+		costComponents = append(costComponents, sharedSQLInstance(tier, availabilityType, dbType, region))
+	} else if sqlInstanceTierToResourceGroup(tier) == "" && strings.Contains(tier, "db-custom-") {
+		cpu, _ := strconv.ParseInt(strings.Split(tier, "-")[2], 10, 32)
+		vCPU := decimalPtr(decimal.NewFromInt32(int32(cpu)))
+
+		costComponents = append(costComponents, cpuCostComponent(region, tier, availabilityType, dbType, vCPU))
+
+		ram, _ := strconv.ParseInt(strings.Split(tier, "-")[3], 10, 32)
+		memory := decimalPtr(decimal.NewFromInt32(int32(ram)).Div(decimal.NewFromInt(1024)))
+
+		costComponents = append(costComponents, r.memoryCostComponent(availabilityType, dbType, memory))
+	} else if strings.Contains(tier, "db-n1-") && dbType == MySQL {
+		costComponents = append(costComponents, sharedSQLInstance(tier, availabilityType, dbType, region))
+	}
+
+	costComponents = append(costComponents, sqlInstanceStorage(region, dbType, availabilityType, diskType, diskSizeGB))
+
+	if !replica {
+		var backupGB *decimal.Decimal
+		if r.BackupStorageGB != nil {
+			backupGB = decimalPtr(decimal.NewFromFloat(*r.BackupStorageGB))
+		}
+		costComponents = append(costComponents, backupCostComponent(region, backupGB))
+
+		if r.UseIPV4 {
+			costComponents = append(costComponents, ipv4CostComponent())
+		}
+	}
+
+	return &schema.Resource{
+		Name:           name,
+		CostComponents: costComponents,
+	}
+}
+
+func (r *SQLDatabaseInstance) memoryCostComponent(availabilityType string, dbType SQLInstanceDBType, memory *decimal.Decimal) *schema.CostComponent {
+	availabilityType = availabilityTypeDescName(availabilityType)
+	dbTypeName := sqlInstanceTypeToDescriptionName(dbType)
+	description := fmt.Sprintf("/%s: %s - RAM/", dbTypeName, availabilityType)
+
+	return &schema.CostComponent{
+		Name:           fmt.Sprintf("Memory (%s)", strings.ToLower(availabilityType)),
+		Unit:           "GB",
+		UnitMultiplier: decimal.NewFromInt(1),
+		HourlyQuantity: memory,
+		ProductFilter: &schema.ProductFilter{
+			VendorName:    strPtr("gcp"),
+			Region:        strPtr(r.Region),
+			Service:       strPtr("Cloud SQL"),
+			ProductFamily: strPtr("ApplicationServices"),
+			AttributeFilters: []*schema.AttributeFilter{
+				{Key: "description", ValueRegex: strPtr(description)},
+			},
+		},
+	}
+}
+
+func cpuCostComponent(region string, tier string, availabilityType string, dbType SQLInstanceDBType, vCPU *decimal.Decimal) *schema.CostComponent {
```

Comment:
I think this function and all the functions below should be methods on the struct. That way they don't pollute the package namespace and we don't need to pass `region`, `tier` and `availabilityType` as args.

Reply:
Applied